File: java/1042.0_Flower_Planting_With_No_Adjacent.java
/*
执行用时：11 ms, 在所有 Java 提交中击败了67.20% 的用户
内存消耗：51 MB, 在所有 Java 提交中击败了17.21% 的用户
通过测试用例：51 / 51
*/
class Solution {
    public int[] gardenNoAdj(int n, int[][] paths) {
        List<Integer> g[] = new ArrayList[n];
        Arrays.setAll(g, index -> new ArrayList<>());
        for (var e: paths) {
            int x = e[0] - 1, y = e[1] - 1;
            g[x].add(y);
            g[y].add(x);
        }
        var color = new int[n];
        for (int i = 0; i < n; ++i) {
            int mask = 1;
            for (var j : g[i]) {
                mask |= 1 << color[j];
            }
            color[i] = Integer.numberOfTrailingZeros(~mask);
        }
        return color;
    }
}


/*
执行用时：16 ms, 在所有 Java 提交中击败了51.08% 的用户
内存消耗：50.1 MB, 在所有 Java 提交中击败了47.31% 的用户
通过测试用例：51 / 51
*/
class Solution {
    public int[] gardenNoAdj(int n, int[][] paths) {
        List<Integer> g[] = new ArrayList[n];
        Arrays.parallelSetAll(g, index -> new ArrayList<>());
        for (var e: paths) {
            int x = e[0] - 1, y = e[1] - 1;
            g[x].add(y);
            g[y].add(x);
        }
        var color = new int[n];
        for (int i = 0; i < n; ++i) {
            int mask = 1;
            for (var j : g[i]) {
                mask |= 1 << color[j];
            }
            color[i] = Integer.numberOfTrailingZeros(~mask);
        }
        return color;
    }
}

File: go/1042.0_Flower_Planting_With_No_Adjacent.go
/*
执行用时：48 ms, 在所有 Go 提交中击败了95.83% 的用户
内存消耗：7.3 MB, 在所有 Go 提交中击败了79.17% 的用户
通过测试用例：51 / 51
*/
func gardenNoAdj(n int, paths [][]int) []int {
    g := make([][]int, n)
    for _, e := range paths {
        x, y := e[0] - 1, e[1] - 1
        g[x] = append(g[x], y)
        g[y] = append(g[y], x)
    }
    color := make([]int, n)
    for i, nodes := range g {
        mask := uint8(1)
        for _, j := range nodes {
            mask |= 1 << color[j]
        }
        color[i] = bits.TrailingZeros8(^mask)
    }
    return color
}
